File: app/src/main/java/com/qwe7002/telegram_rc/static_class/sms.java
package com.qwe7002.telegram_rc.static_class;

import android.Manifest;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.telephony.SmsManager;

import androidx.core.content.PermissionChecker;

import com.google.gson.Gson;
import com.qwe7002.telegram_rc.R;
import com.qwe7002.telegram_rc.data_structure.request_message;
import com.qwe7002.telegram_rc.sms_send_receiver;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Objects;

import okhttp3.Call;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class sms {
    public static void send_fallback_sms(Context context, String content, int sub_id) {
        final String TAG = "send_fallback_sms";
        if (androidx.core.content.PermissionChecker.checkSelfPermission(context, Manifest.permission.SEND_SMS) != PermissionChecker.PERMISSION_GRANTED) {
            android.util.Log.d(TAG, ": No permission.");
            return;
        }
        SharedPreferences sharedPreferences = context.getSharedPreferences("data", Context.MODE_PRIVATE);
        String trust_number = sharedPreferences.getString("trusted_phone_number", null);
        if (trust_number == null) {
            android.util.Log.i(TAG, "The trusted number is empty.");
            return;
        }
        if (!sharedPreferences.getBoolean("fallback_sms", false)) {
            android.util.Log.i(TAG, "Did not open the SMS to fall back.");
            return;
        }
        android.telephony.SmsManager sms_manager;
        if (sub_id == -1) {
            sms_manager = SmsManager.getDefault();
        } else {
            sms_manager = SmsManager.getSmsManagerForSubscriptionId(sub_id);
        }
        ArrayList<String> divideContents = sms_manager.divideMessage(content);
        sms_manager.sendMultipartTextMessage(trust_number, null, divideContents, null, null);
    }

    public static void send_sms(Context context, String send_to, String content, int slot, int sub_id) {
        send_sms(context, send_to, content, slot, sub_id, -1);
    }

    public static void send_sms(Context context, String send_to, String content, int slot, int sub_id, long message_id) {
        if (PermissionChecker.checkSelfPermission(context, Manifest.permission.SEND_SMS) != PermissionChecker.PERMISSION_GRANTED) {
            android.util.Log.d("send_sms", "No permission.");
            return;
        }
        if (!other.isPhoneNumber(send_to)) {
            log.writeLog(context, "[" + send_to + "] is an illegal phone number");
            return;
        }
        SharedPreferences sharedPreferences = context.getSharedPreferences("data", Context.MODE_PRIVATE);
        String bot_token = sharedPreferences.getString("bot_token", "");
        String chat_id = sharedPreferences.getString("chat_id", "");
        String request_uri = network.getUrl(bot_token, "sendMessage");
        if (message_id != -1) {
            android.util.Log.d("send_sms", "Find the message_id and switch to edit mode.");
            request_uri = network.getUrl(bot_token, "editMessageText");
        }
        request_message request_body = new request_message();
        request_body.chat_id = chat_id;
        SmsManager sms_manager;
        if (sub_id == -1) {
            sms_manager = SmsManager.getDefault();
        } else {
            sms_manager = SmsManager.getSmsManagerForSubscriptionId(sub_id);
        }
        String dual_sim = other.getDualSimCardDisplay(context, slot, sharedPreferences.getBoolean("display_dual_sim_display_name", false));
        String send_content = "[" + dual_sim + context.getString(R.string.send_sms_head) + "]" + "\n" + context.getString(R.string.to) + send_to + "\n" + context.getString(R.string.content) + content;
        request_body.text = send_content + "\n" + context.getString(R.string.status) + context.getString(R.string.sending);
        request_body.message_id = message_id;
        Gson gson = new Gson();
        String request_body_raw = gson.toJson(request_body);
        RequestBody body = RequestBody.create(request_body_raw, CONST.JSON);
        OkHttpClient okhttp_client = network.getOkhttpObj(sharedPreferences.getBoolean("doh_switch", true));
        Request request = new Request.Builder().url(request_uri).method("POST", body).build();
        Call call = okhttp_client.newCall(request);
        try {
            Response response = call.execute();
            if (response.code() != 200 || response.body() == null) {
                throw new IOException(String.valueOf(response.code()));
            }
            if (message_id == -1) {
                message_id = other.getMessageId(Objects.requireNonNull(response.body()).string());
            }
        } catch (IOException e) {
            e.printStackTrace();
            log.writeLog(context, "failed to send message:" + e.getMessage());
        }
        ArrayList<String> divideContents = sms_manager.divideMessage(content);
        ArrayList<PendingIntent> send_receiver_list = new ArrayList<>();
        IntentFilter filter = new IntentFilter("send_sms");
        BroadcastReceiver receiver = new sms_send_receiver();
        context.getApplicationContext().registerReceiver(receiver, filter);
        Intent sent_intent = new Intent("send_sms");
        sent_intent.putExtra("message_id", message_id);
        sent_intent.putExtra("message_text", send_content);
        sent_intent.putExtra("sub_id", sms_manager.getSubscriptionId());
        PendingIntent sentIntent = PendingIntent.getBroadcast(context, 0, sent_intent, PendingIntent.FLAG_CANCEL_CURRENT);
        send_receiver_list.add(sentIntent);
        sms_manager.sendMultipartTextMessage(send_to, null, divideContents, send_receiver_list, null);
    }
}

File: app/src/main/java/com/qwe7002/telegram_rc/static_class/SMS.kt
@file:Suppress("DEPRECATION")

package com.qwe7002.telegram_rc.static_class

import android.Manifest
import android.annotation.SuppressLint
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.telephony.SmsManager
import android.util.Log
import androidx.core.content.PermissionChecker
import com.google.gson.Gson
import com.qwe7002.telegram_rc.R
import com.qwe7002.telegram_rc.SMSSendResultReceiver
import com.qwe7002.telegram_rc.data_structure.RequestMessage
import com.qwe7002.telegram_rc.static_class.LogManage.writeLog
import com.tencent.mmkv.MMKV
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.Objects

object SMS {
    @JvmStatic
    fun sendFallbackSMS(context: Context, content: String?, subId: Int) {
        val TAG = "send_fallback_sms"
        if (PermissionChecker.checkSelfPermission(
                context,
                Manifest.permission.SEND_SMS
            ) != PermissionChecker.PERMISSION_GRANTED
        ) {
            Log.d(TAG, ": No permission.")
            return
        }
        val preferences = MMKV.defaultMMKV()
        val trustNumber = preferences.getString("trusted_phone_number", "")
        if (trustNumber.isNullOrEmpty()) {
            Log.i(TAG, "The trusted number is empty.")
            return
        }
        if (!preferences.getBoolean("fallback_sms", false)) {
            Log.i(TAG, "Did not open the SMS to fall back.")
            return
        }
        if (content.isNullOrEmpty()) {
            Log.i(TAG, "The content is empty.")
            return
        }
        try {
            val smsManager = if (subId == -1) {
                SmsManager.getDefault()
            } else {
                SmsManager.getSmsManagerForSubscriptionId(subId)
            }
            val divideContents = smsManager.divideMessage(content)
            smsManager.sendMultipartTextMessage(trustNumber, null, divideContents, null, null)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send fallback SMS: ${e.message}")
        }
    }

    @JvmStatic
    fun sendSMS(context: Context, sendTo: String, content: String, slot: Int, subId: Int) {
        sendSMS(context, sendTo, content, slot, subId, -1)
    }

    @JvmStatic
    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    fun sendSMS(
        context: Context,
        sendTo: String,
        content: String,
        slot: Int,
        subId: Int,
        messageId: Long
    ) {
        var privateMessageId = messageId
        if (PermissionChecker.checkSelfPermission(
                context,
                Manifest.permission.SEND_SMS
            ) != PermissionChecker.PERMISSION_GRANTED
        ) {
            Log.d("send_sms", "No permission.")
            return
        }
        if (!Other.isPhoneNumber(sendTo)) {
            writeLog(context, "[$sendTo] is an illegal phone number")
            return
        }
        val preferences = MMKV.defaultMMKV()
        val botToken = preferences.getString("bot_token", "") ?: ""
        val chatId = preferences.getString("chat_id", "") ?: ""
        var requestUri = Network.getUrl(botToken, "sendMessage")
        if (privateMessageId != -1L) {
            Log.d("send_sms", "Find the message_id and switch to edit mode.")
            requestUri = Network.getUrl(botToken, "editMessageText")
        }
        val requestBody = RequestMessage()
        requestBody.chatId = chatId
        val smsManager = if (subId == -1) {
            SmsManager.getDefault()
        } else {
            SmsManager.getSmsManagerForSubscriptionId(subId)
        }
        val dualSim = Other.getDualSimCardDisplay(
            context,
            slot,
            preferences.getBoolean("display_dual_sim_display_name", false)
        )
        val sendContent = "[$dualSim${context.getString(R.string.send_sms_head)}]\n${context.getString(R.string.to)}$sendTo\n${context.getString(R.string.content)}$content"
        requestBody.text = "$sendContent\n${context.getString(R.string.status)}${context.getString(R.string.sending)}"
        requestBody.messageId = privateMessageId
        val gson = Gson()
        val requestBodyRaw = gson.toJson(requestBody)
        val body: RequestBody = requestBodyRaw.toRequestBody(Const.JSON)
        val okhttpClient = Network.getOkhttpObj()
        val request: Request = Request.Builder().url(requestUri).method("POST", body).build()
        val call = okhttpClient.newCall(request)
        try {
            val response = call.execute()
            if (response.code != 200) {
                writeLog(context, "Failed to send message: HTTP ${response.code}")
                response.close()
                return
            }
            if (privateMessageId == -1L) {
                privateMessageId = Other.getMessageId(Objects.requireNonNull(response.body).string())
            }
            response.close()
        } catch (e: Exception) {
            Log.d("sendSMS", "sendSMS: $e")
            writeLog(context, "failed to send message:" + e.message)
            return
        }
        val divideContents = smsManager.divideMessage(content)
        val sendReceiverList = ArrayList<PendingIntent>()
        val filter = IntentFilter("send_sms")
        val receiver: BroadcastReceiver = SMSSendResultReceiver()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            context.applicationContext.registerReceiver(receiver, filter, Context.RECEIVER_EXPORTED)
        } else {
            context.applicationContext.registerReceiver(receiver, filter)
        }
        val intent = Intent("send_sms")
        intent.putExtra("message_id", privateMessageId)
        intent.putExtra("message_text", sendContent)
        intent.putExtra("sub_id", smsManager.subscriptionId)
        val sentIntent = PendingIntent.getBroadcast(
            context,
            privateMessageId.toInt(),
            intent,
            PendingIntent.FLAG_CANCEL_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        sendReceiverList.add(sentIntent)
        try {
            smsManager.sendMultipartTextMessage(
                sendTo,
                null,
                divideContents,
                sendReceiverList,
                null
            )
        } catch (e: Exception) {
            Log.e("sendSMS", "Failed to send SMS: ${e.message}")
            writeLog(context, "Failed to send SMS: ${e.message}")
            context.applicationContext.unregisterReceiver(receiver)
        }
    }
}
